chain/chain/src/store.rs

```diff
@@ -1815,116 +1835,211 @@ impl<'a> ChainStoreUpdate<'a> {
+            self.chain_store
+                .incoming_receipts
+                .cache_remove(&get_block_shard_id(&block_hash, shard_id));
+            // 2c. Delete from chunk_hash_per_height_shard (ColChunkPerHeightShard)
+            store_update.delete(ColChunkPerHeightShard, &get_height_shard_id(height, shard_id));
+            self.chain_store
+                .chunk_hash_per_height_shard
+                .cache_remove(&get_height_shard_id(height, shard_id));
+            // 2d. Delete from next_block_with_new_chunk (ColNextBlockWithNewChunk)
+            store_update
+                .delete(ColNextBlockWithNewChunk, &get_block_shard_id(&block_hash, shard_id));
+            self.chain_store
+                .next_block_with_new_chunk
+                .cache_remove(&get_block_shard_id(&block_hash, shard_id));
+            // 2e. Delete from ColStateHeaders
+            let key = StateHeaderKey(shard_id, block_hash).try_to_vec()?;
+            store_update.delete(ColStateHeaders, &key);
+            // 2f. Delete from ColStateParts
+            // Already done, check chain.clear_downloaded_parts()
+        }
+        for chunk_header in block.chunks {
+            let (receipts, transactions) = match self.get_chunk_clone_from_header(&chunk_header) {
+                Ok(chunk) => (chunk.receipts, chunk.transactions),
+                _ => (vec![], vec![]),
+            };
+            // 2g. Delete from receipt_id_to_shard_id (ColReceiptIdToShardId)
+            for receipt in receipts {
+                store_update.delete(ColReceiptIdToShardId, receipt.receipt_id.as_ref());
+                self.chain_store.receipt_id_to_shard_id.cache_remove(&receipt.receipt_id.into());
+            }
+            // 2h. Delete from ColTransactions
+            for transaction in transactions {
+                store_update.delete(ColTransactions, transaction.get_hash().as_ref());
+                self.chain_store.transactions.cache_remove(&transaction.get_hash().into());
+            }
+
+            // 3. Delete chunk_hash-indexed data
+            // 3a. Delete chunks (ColChunks)
+            store_update.delete(ColChunks, chunk_header.hash.as_ref());
+            self.chain_store.chunks.cache_remove(&chunk_header.hash.clone().into());
+            // 3b. Delete chunk extras (ColChunkExtra)
+            store_update.delete(ColChunkExtra, chunk_header.hash.as_ref());
+            self.chain_store.chunk_extras.cache_remove(&chunk_header.hash.clone().into());
+            // 3c. Delete partial_chunks (ColPartialChunks)
+            store_update.delete(ColPartialChunks, chunk_header.hash.as_ref());
+            self.chain_store.partial_chunks.cache_remove(&chunk_header.hash.clone().into());
+            // 3d. Delete invalid chunks (ColInvalidChunks)
+            store_update.delete(ColInvalidChunks, chunk_header.hash.as_ref());
+            self.chain_store.invalid_chunks.cache_remove(&chunk_header.hash.clone().into());
```

Comment:
1. Let's clone and convert the hash only once instead of 4 times.
2. It seems that `.into()` is enough, you don't need to `.clone()`

Apply the same for the `block_hash` below.